micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java

```diff
@@ -121,50 +122,75 @@ public CaffeineCacheMetrics(Cache<?, ?> cache, String cacheName, Iterable<Tag> t
 
     @Override
     protected Long size() {
-        return cache.estimatedSize();
+        Cache<?, ?> ref = cache.get();
+        if (ref != null) {
+            return ref.estimatedSize();
+        }
+
+        return null;
     }
 
     @Override
     protected long hitCount() {
-        return cache.stats().hitCount();
+        Cache<?, ?> ref = cache.get();
+        if (ref != null) {
+            return ref.stats().hitCount();
+        }
+
+        return 0L;
     }
 
     @Override
     protected Long missCount() {
-        return cache.stats().missCount();
+        Cache<?, ?> ref = cache.get();
+        if (ref != null) {
+            return ref.stats().missCount();
+        }
+
+        return null;
     }
 
     @Override
     protected Long evictionCount() {
-        return cache.stats().evictionCount();
+        Cache<?, ?> ref = cache.get();
+        if (ref != null) {
+            return ref.stats().evictionCount();
+        }
+
+        return null;
     }
 
     @Override
     protected long putCount() {
-        return cache.stats().loadCount();
+        Cache<?, ?> ref = cache.get();
+        if (ref != null) {
+            return ref.stats().loadCount();
+        }
+
+        return 0L;
     }
 
     @Override
     protected void bindImplementationSpecificMetrics(MeterRegistry registry) {
-        FunctionCounter.builder("cache.eviction.weight", cache, c -> c.stats().evictionWeight())
+        FunctionCounter.builder("cache.eviction.weight", cache.get(), c -> c.stats().evictionWeight())
                 .tags(getTagsWithCacheName())
                 .description("The sum of weights of evicted entries. This total does not include manual invalidations.")
                 .register(registry);
 
         if (cache instanceof LoadingCache) {
```

Comment:
This is why the test is failing. The type needs to be checked on `cache.get()` now

Reply:
Ah, it's so obvious once you see it. Thanks!

Reply:
Fixed, and the tests are green :tada: